packages/components/src/date-time/test/date.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+import { shallow } from 'enzyme';
+import moment from 'moment';
+
+/**
+ * Internal dependencies
+ */
+import DatePicker from '../date';
+
+const TIMEZONELESS_FORMAT = 'YYYY-MM-DDTHH:mm:ss';
+
+describe( 'DatePicker', () => {
+	it( 'should pass down a moment object for currentDate', () => {
+		const currentDate = '1986-10-18T23:00:00';
+		const wrapper = shallow( <DatePicker currentDate={ currentDate } /> );
+		const date = wrapper.children().props().date;
+		expect( moment.isMoment( date ) ).toBe( true );
+		expect( date.isSame( moment( currentDate ) ) ).toBe( true );
+	} );
+
+	it( 'should pass down a null date when currentDate is set to null', () => {
+		const wrapper = shallow( <DatePicker currentDate={ null } /> );
+		expect( wrapper.children().props().date ).toBeNull();
+	} );
+
+	it( 'should pass down a moment object for now when currentDate is undefined', () => {
+		const wrapper = shallow( <DatePicker /> );
+		const date = wrapper.children().props().date;
+		expect( moment.isMoment( date ) ).toBe( true );
+		expect( date.isSame( moment(), 'second' ) ).toBe( true );
+	} );
+
+	describe( 'getMomentDate', () => {
+		it( 'should return a Moment object representing a given date string', () => {
+			const currentDate = '1986-10-18T23:00:00';
+			const wrapper = shallow( <DatePicker /> );
+			const momentDate = wrapper.instance().getMomentDate( currentDate );
+
+			expect( moment.isMoment( momentDate ) ).toBe( true );
+			expect( momentDate.isSame( moment( currentDate ) ) ).toBe( true );
+		} );
+
+		it( 'should return null when given a null agrument', () => {
+			const currentDate = null;
+			const wrapper = shallow( <DatePicker /> );
+			const momentDate = wrapper.instance().getMomentDate( currentDate );
+
+			expect( momentDate ).toBeNull();
+		} );
+
+		it( 'should return a Moment object representing now when given an undefined argument', () => {
+			const wrapper = shallow( <DatePicker /> );
+			const momentDate = wrapper.instance().getMomentDate();
+
+			expect( moment.isMoment( momentDate ) ).toBe( true );
+			expect( momentDate.isSame( moment(), 'second' ) ).toBe( true );
+		} );
+	} );
+
+	describe( 'onChangeMoment', () => {
+		it( 'should call onChange with a formated date of the input', () => {
+			const onChangeSpy = jest.fn();
+			const currentDate = '1986-10-18T11:00:00';
+			const wrapper = shallow( <DatePicker currentDate={ currentDate } onChange={ onChangeSpy } /> );
+			const newDate = moment();
+
+			wrapper.instance().onChangeMoment( newDate );
+
+			expect( onChangeSpy ).toHaveBeenCalledWith( newDate.format( TIMEZONELESS_FORMAT ) );
+		} );
+
+		it( 'should call onChange with hours, minutes, seconds of the current time when currentDate is undefined', () => {
+			const onChangeSpy = jest.fn();
+			const wrapper = shallow( <DatePicker onChange={ onChangeSpy } /> );
+			const newDate = moment( '1986-10-18T11:00:00' );
+			const current = moment();
```

Comment:
This worries me a bit; I feel like it will be prone to test fragility, where the time required to execute the test is such that `moment()` called here will produce a different time than `moment()` called within the implementation of `onChangeMoment`. Usually this should take a mere fraction of a second, but I've witnessed in the past where it's enough time to tick into the next second and cause an uenxpected test failure.

I think our options would be:

- Find a way to provide or otherwise control the current time as referenced within `onChangeMoment`
- Mock `moment` to have full control over what it returns†
- Provide some flexibility in the assertion (i.e. within a +/- allowable timeframe)
- Don't test it

Reply:
Lets go with `Provide some flexibility in the assertion (i.e. within a +/- allowable timeframe)` for $1000 please, Alex.

I figured comparison on basis of seconds was enough to overcome differences in execution time, but in an effort to prevent future headaches, a change is a good idea.

I changed the tests to save the argument passed to onChange for later comparison using moment's `isSame` function. Then I compare with a granularity of `minute`, which would preclude all but the rarest of circumstances where a pause in execution time ticks into the next second AND minute.